### ISSUE TITLE
`dub test` succeeds but `dub build` fails on same project

### ISSUE THREAD

**Issue by avnomad** [dlang/dub#1981]
### System information
- **dub version**: DUB version 1.21.0, built on May 11 2020
- **OS Platform and distribution**: Linux Mint 18 (Sarah)
- **compiler version**: DMD64 D Compiler v2.092.0

### Bug Description

It's is possible for the `dub test` command to succeed even if the module containing the main function of an application  had compiler or linker errors. Such errors then manifest if you attempt to call `dub build` or `dub run`.

### How to reproduce?

- Start with an empty directory.
- `dub init` (use default parameters)
- Open `app.d` and remove the closing brace `}`.
- Create and empty `test.d` file in the `source` directory.
- From the directory that hosts `dub.sdl` do `dub test`. You'll notice that it succeeds while it shouldn't.
- From the same directory do `dub build`. You'll see an error like:
    ```
    Performing "debug" build using /usr/bin/dmd for x86_64.
    bug ~master: building configuration "application"...
    source/app.d(7,1): Error: found End of File when expecting } following compound statement
    /usr/bin/dmd failed with exit code 1.
    ````

### Expected Behaviour

Maybe I'm influenced by how `maven` does things for Java, but I believe `dub test` should imply `dub build`. I don't think it makes any sense to start running tests if your application can't even compile.

### Workarounds

Whenever you'd call `dub test`, call `dub build && dub test` instead.

**Comment by andre2007** [comment]
Maybe as explanation why this happens. Afaik dub test will not use your app.d by default but generates an own app.d.

If you want to use the app.d from your project you can use the attribute `mainSourceFile`.

**Comment by Geod24** [comment]
@avnomad : I would say that is by design, and not a bug. `unittest` is inherently a different "mode" than regular build, so you are bound to see differences. The exclusion of the "main" source file is the most common one, and perhaps with the new mode in druntime that doesn't run the `main` code after unittests we could make sure it compiles, but then one day someone would hit the second most common case, having something that is in a `version (unittest)` used in their production code and open another issue, and the cycle would repeat.

**Comment by Geod24** [comment]
Closing for lack of activity. As explained, this is by design. It's a bit more apparent now that dub excludes other configuration's `mainSourceFile`.

**Comment by avnomad** [comment]
Hello @Geod24 and @andre2007. I'm really sorry for not responding in time. It seems something is really wrong with my notification settings because I never got an e-mail and only got a notification in the UI when you closed the issue.

@andre2007 I'm aware dub doesn't use the main from my `app.d` but generates it's own main function. That's so that my main application isn't run. But that doesn't imply that the `app.d` file shouldn't pass through the compiler or even the linker. I'm afraid the `mainSourceFile` attribute does the opposite of what I want: it doesn't _include_ the given file for compilation when doing a `dub test`: it _excludes_ it. I tried adding the line `mainSourceFile "source/app.d"` in `dub.sdl` in the example above and the problem persists.

@Geod24 I don't feel very strongly about calling this a bug vs 'by design', but you have to agree this is bad user experience and not consistent with how similar tools in the industry, including IDEs, work. Again, there is nothing wrong with carving your own path, but in this case I don't see the benefit.

A change in the D runtime doesn't seem necessary: you can keep generating you own main, compiling it and linking it with the test code. You'll also compile and link the application's main. In effect you'll create two executables, but only run the one which runs the tests. That should be a lot faster than running `dub` twice and won't leave people thinking their application compiles while it doesn't.

Say an application named `app` has two files: `app.d` and `impl.d`. Your dependency graph for `dub test` looks like:
```
dub-generated-main.d -> dub-generated-main.o
impl.d -> impl.o
dub-generated-main.o, impl.o -> app-test-library (executable)
```
While for `dub build` it looks like:
```
app.d -> app.o
impl.d -> impl.o
app.o, impl.o -> app (executable)
```
I'm proposing combining the two so now the dependency graph for `dub test` is:
```
dub-generated-main.d -> dub-generated-main.o
impl.d -> impl.o
dub-generated-main.o, impl.o -> app-test-library (executable)
app.d -> app.o
app.o, impl.o -> app (executable)
```
Once you are done compiling and linking, you'll still run `app-test-library` as you do today.

About `version (unittest)` I don't think anyone expects a single build to check both paths any more than for `#if #else #endif` in C/C++. But as for `#if #else #endif` blocks in C/C++, I believe people are expecting the syntax of the conditional compilation statements themselves to be checked. The way things are right now you can have a `version (unittest)` block without a closing brace and `dub test` will not complain. (Also see point 3 in [D's Conditional Compilation documentation](https://dlang.org/spec/version.html) "Any DeclarationBlock or Statement that is not compiled in still must be syntactically correct.")

**Comment by PetarKirov** [comment]
I agree with @avnomad that the user experience of `dub test` is suboptimal. IMO we should either change `dub test` to also build `app.d` or add new mode, say `dub ci` (see e.g. https://docs.npmjs.com/cli/v6/commands/npm-ci) that is a "full" build + test. 

Such a change could be a good candidate for Dub 2.0 - https://github.com/dlang/dub/discussions/2069
